Fix crash on Android 9

Rafael2616 opened the pull request:
Fixes #3778 <!-- optional -->

## Type of change
<!-- Replace :x: with :white_check_mark: to "check" the specified bullet -->

:x: General change (non-breaking change that doesn't fit the below categories like copyediting)
:white_check_mark: Bug fix (non-breaking change which fixes an issue)
:x: New feature (non-breaking change which adds functionality)
:x: Breaking change (fix or feature that would cause existing functionality to not work as expected)

Rafael2616 committed:
fix #3778

Rafael2616 committed:
Fix Java syntax

Rafael2616 committed:
Merge branch 'LawnchairLauncher:13-dev' into 13-dev

Rafael2616 committed:
Update build_debug_apk.yml

Rafael2616 committed:
Merge branch 'LawnchairLauncher:13-dev' into 13-dev

Rafael2616 committed:
Update build_debug_apk.yml

Rafael2616 committed:
fix for #3778

MrSluffy commented:
There's some crashes still 

```
Caused by: java.lang.NoClassDefFoundError: Failed resolution of: Lcom/android/internal/policy/ScreenDecorationsUtils;
	at com.android.systemui.shared.system.QuickStepContract.supportsRoundedCornersOnWindows(QuickStepContract.java:384)
	at app.lawnchair.util.LawnchairUtilsKt.supportsRoundedCornersOnWindows(LawnchairUtils.kt:127)
	at com.android.quickstep.util.TaskCornerRadius.get(TaskCornerRadius.java:30)
	at com.android.quickstep.views.TaskView$FullscreenDrawParams.<init>(TaskView.java:1773)
	at com.android.quickstep.views.TaskView.<init>(TaskView.java:431)
	at com.android.quickstep.views.TaskView.<init>(TaskView.java:422)
	at com.android.quickstep.views.TaskView.<init>(TaskView.java:418)
	... 12 more
Caused by: java.lang.ClassNotFoundException: Didn't find class "com.android.internal.policy.ScreenDecorationsUtils" on path: DexPathList[[zip file "/data/app/app.lawnchair.debug-2IYx9NyOOjIFsbEGavGNSw==/base.apk"],nativeLibraryDirectories=[/data/app/app.lawnchair.debug-2IYx9NyOOjIFsbEGavGNSw==/lib/arm, /data/app/app.lawnchair.debug-2IYx9NyOOjIFsbEGavGNSw==/base.apk!/lib/armeabi-v7a, /system/lib, /system/product/lib]]
	at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:134)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
```

You can try add condition here atleastq too [`supportsRoundedCornersOnWindows`](https://github.com/LawnchairLauncher/lawnchair/blob/13-dev/lawnchair/src/app/lawnchair/util/LawnchairUtils.kt#L123)

Rafael2616 commented:
> There's some crashes still 
> 
> ```
> Caused by: java.lang.NoClassDefFoundError: Failed resolution of: Lcom/android/internal/policy/ScreenDecorationsUtils;
> 	at com.android.systemui.shared.system.QuickStepContract.supportsRoundedCornersOnWindows(QuickStepContract.java:384)
> 	at app.lawnchair.util.LawnchairUtilsKt.supportsRoundedCornersOnWindows(LawnchairUtils.kt:127)
> 	at com.android.quickstep.util.TaskCornerRadius.get(TaskCornerRadius.java:30)
> 	at com.android.quickstep.views.TaskView$FullscreenDrawParams.<init>(TaskView.java:1773)
> 	at com.android.quickstep.views.TaskView.<init>(TaskView.java:431)
> 	at com.android.quickstep.views.TaskView.<init>(TaskView.java:422)
> 	at com.android.quickstep.views.TaskView.<init>(TaskView.java:418)
> 	... 12 more
> Caused by: java.lang.ClassNotFoundException: Didn't find class "com.android.internal.policy.ScreenDecorationsUtils" on path: DexPathList[[zip file "/data/app/app.lawnchair.debug-2IYx9NyOOjIFsbEGavGNSw==/base.apk"],nativeLibraryDirectories=[/data/app/app.lawnchair.debug-2IYx9NyOOjIFsbEGavGNSw==/lib/arm, /data/app/app.lawnchair.debug-2IYx9NyOOjIFsbEGavGNSw==/base.apk!/lib/armeabi-v7a, /system/lib, /system/product/lib]]
> 	at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:134)
> 	at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
> 	at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
> ```
> 
> You can try add condition here atleastq too [`supportsRoundedCornersOnWindows`](https://github.com/LawnchairLauncher/lawnchair/blob/13-dev/lawnchair/src/app/lawnchair/util/LawnchairUtils.kt#L123)

Wow, that was a good start, 3 problems fixed with the same fix. My record

MrSluffy commented:
> What value should I put here for below Q, true or false?
> 
> ```
>     return QuickStepContract.supportsRoundedCornersOnWindows(context.resources)
> ```

false

Rafael2616 committed:
more fixes

Rafael2616 commented:
> > What value should I put here for below Q, true or false?
> > 
> > ```
> >     return QuickStepContract.supportsRoundedCornersOnWindows(context.resources)
> > ```
> 
> false
> 

Done 👍

MrSluffy commented:
> > > What value should I put here for below Q, true or false?
> > > ```
> > >     return QuickStepContract.supportsRoundedCornersOnWindows(context.resources)
> > > ```
> > 
> > 
> > false
> 
> Done 👍

No, use this intead and test it
```
fun getWindowCornerRadius(context: Context): Float {
    val prefs = getPrefsIfUnlocked(context)
    return when {
        prefs?.overrideWindowCornerRadius?.get() == true -> prefs.windowCornerRadius.get().toFloat()
        Utilities.ATLEAST_Q -> QuickStepContract.getWindowCornerRadius(context)
        else -> 0.0f
    }
}
```

Rafael2616 commented:
Sorry, I changed the wrong function

Rafael2616 commented:
I will correct it now

Rafael2616 committed:
Update LawnchairUtils.kt

Rafael2616 committed:
fix more issue

Goooler committed:
Add a switch to control suggested apps showing (#3803)

Rafael2616 commented:
Done again

MrSluffy commented:
after build you can test

MrSluffy commented:
> Done again

this too : 

```
fun supportsRoundedCornersOnWindows(context: Context): Boolean {
    val prefs = getPrefsIfUnlocked(context)

    return when {
        prefs?.overrideWindowCornerRadius?.get() == true -> true
        Utilities.ATLEAST_Q -> QuickStepContract.supportsRoundedCornersOnWindows(context.resources)
        else -> false
    }
}
```

MrSluffy committed:
Fixed crash oppo A11 (#3809)

logs : https://katb.in/icicokoxuru

MrSluffy committed:
Fixed crash on a11 oneui (#3801)

android.view.InflateException: Binary XML file line #27 in app.lawnchair.debug:layout/task_desktop: Resource

MrSluffy committed:
another fixed oneui 11 (#3810)

android.view.InflateException: Binary XML file line #25 in app.lawnchair.debug:layout/task_desktop: Resource ID #0x106002f
Caused by: android.content.res.Resources$NotFoundException: Resource ID #0x106002f
	at android.content.res.ResourcesImpl.getValue(ResourcesImpl.java:255)
	at android.content.res.Resources.getColor(Resources.java:1169)
	at com.android.quickstep.views.DesktopTaskView.onFinishInflate(DesktopTaskView.java:124)
	at android.view.LayoutInflater.rInflate(LayoutInflater.java:1134)
	at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:1082)
	at android.view.LayoutInflater.inflate(LayoutInflater.java:680)
	at android.view.LayoutInflater.inflate(LayoutInflater.java:532)
	at com.android.launcher3.util.ViewPool.inflateNewView(ViewPool.java:106)
	at com.android.launcher3.util.ViewPool.lambda$initPool$1(ViewPool.java:69)
	at com.android.launcher3.util.ViewPool.$r8$lambda$wSRnQcMZiOmM7xwf4gm88rNehhg(Unknown Source:0)
	at com.android.launcher3.util.ViewPool$$ExternalSyntheticLambda1.run(Unknown Source:8)
	at java.lang.Thread.run(Thread.java:923)

Rafael2616 committed:
Apply suggestions in LawnchairUtils.kt

Rafael2616 committed:
Try more one fix

Rafael2616 commented:
I limited this function to Android 10 +, maybe it will work now

Rafael2616 committed:
Merge branch 'fix-crash-a9' into Rafael2616-patch-1

MrSluffy committed:
Fix DT2S (Root) using keyevent (#3799)

* Add alternative root execution for dt2s

* Fix styling

* Use Shell instead

Rafael2616 commented:
@MrSluffy Can you review my last commit and run it if it's correct, so I can see if there are any more problems launching the app?

MrSluffy commented:
> Author

I can't test right now, but looks fine to me, test it now

Rafael2616 commented:
More one:

```
Lawnchair (Debug) bug report 1 de dez de 2023 21:26:18
version: 13 Dev (#2478) (13)
commit: 195618a
build.brand: lge
build.device: mmh4
build.display: PKQ1.190522.001
build.fingerprint: lge/mmh4_f/mmh4:9/PKQ1.190522.001/2106722188f70:user/release-keys
build.hardware: mmh4
build.id: PKQ1.190522.001
build.manufacturer: LGE
build.model: LM-X420
build.security.level: 2021-03-01
build.product: mmh4_f
build.type: user
version.codename: REL
version.incremental: 2106722188f70
version.release: 9
version.sdk_int: 28
display.density_dpi: 280
isRecentsEnabled: false

error: Uncaught exception

java.lang.RuntimeException: Unable to start activity ComponentInfo{app.lawnchair.debug/app.lawnchair.LawnchairLauncher}: java.lang.NullPointerException: Attempt to invoke virtual method 'void com.android.systemui.shared.system.SystemGestureExclusionListenerCompat.unregister()' on a null object reference
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3021)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3156)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1864)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loop(Looper.java:205)
	at android.app.ActivityThread.main(ActivityThread.java:6993)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:884)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void com.android.systemui.shared.system.SystemGestureExclusionListenerCompat.unregister()' on a null object reference
	at com.android.quickstep.RecentsAnimationDeviceState.onDisplayInfoChanged(RecentsAnimationDeviceState.java:266)
	at com.android.quickstep.RecentsAnimationDeviceState.<init>(RecentsAnimationDeviceState.java:176)
	at com.android.quickstep.RecentsAnimationDeviceState.<init>(RecentsAnimationDeviceState.java:129)
	at com.android.quickstep.util.SplitWithKeyboardShortcutController.<init>(SplitWithKeyboardShortcutController.java:72)
	at com.android.launcher3.uioverrides.QuickstepLauncher.setupViews(QuickstepLauncher.java:230)
	at app.lawnchair.LawnchairLauncher.setupViews(LawnchairLauncher.kt:272)
	at com.android.launcher3.Launcher.onCreate(Launcher.java:518)
	at com.android.launcher3.uioverrides.QuickstepLauncher.onCreate(QuickstepLauncher.java:527)
	at app.lawnchair.LawnchairLauncher.onCreate(LawnchairLauncher.kt:208)
	at android.app.Activity.performCreate(Activity.java:7159)
	at android.app.Activity.performCreate(Activity.java:7150)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1272)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3001)
	... 11 more

```

Rafael2616 committed:
fix: avoid NPE in mExclusionListener due to it being null under Android 10

Rafael2616 commented:
I added a null check, since the last commit made this `mExclusionListener` null, since it was not initialized (because it is also not supported)

renovate[bot] committed:
Update Dagger to v2.49 (#3811)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

Rafael2616 commented:
I forgot `;` 🫥

Rafael2616 committed:
fix Java sintax

Rafael2616 committed:
More fix

Rafael2616 committed:
Fix prop not initialized

lawnchair-bot committed:
New Crowdin translations by GitHub Action (#3819)

Co-authored-by: Crowdin Bot <support+bot@crowdin.com>

Rafael2616 commented:
I fixed one more problem, now I'm in this

```
Lawnchair (Debug) bug report 2 de dez de 2023 21:21:32
version: 13 Dev (#9) (13)
commit: ee9e679
build.brand: lge
build.device: mmh4
build.display: PKQ1.190522.001
build.fingerprint: lge/mmh4_f/mmh4:9/PKQ1.190522.001/2106722188f70:user/release-keys
build.hardware: mmh4
build.id: PKQ1.190522.001
build.manufacturer: LGE
build.model: LM-X420
build.security.level: 2021-03-01
build.product: mmh4_f
build.type: user
version.codename: REL
version.incremental: 2106722188f70
version.release: 9
version.sdk_int: 28
display.density_dpi: 280
isRecentsEnabled: false

error: Uncaught exception

java.lang.NoSuchMethodError: No virtual method getSourceLayoutResId()I in class Landroid/view/View; or its super classes (declaration of 'android.view.View' appears in /system/framework/framework.jar!classes2.dex)
	at com.android.launcher3.Workspace.setPageIndicatorInset(Workspace.java:412)
	at com.android.launcher3.Workspace.setInsets(Workspace.java:365)
	at com.android.launcher3.InsettableFrameLayout.setFrameLayoutChildInsets(InsettableFrameLayout.java:29)
	at com.android.launcher3.InsettableFrameLayout.setInsets(InsettableFrameLayout.java:44)
	at com.android.launcher3.InsettableFrameLayout.setFrameLayoutChildInsets(InsettableFrameLayout.java:29)
	at com.android.launcher3.InsettableFrameLayout.setInsets(InsettableFrameLayout.java:44)
	at com.android.launcher3.LauncherRootView.dispatchInsets(LauncherRootView.java:87)
	at com.android.launcher3.Launcher.onCreate(Launcher.java:555)
	at com.android.launcher3.uioverrides.QuickstepLauncher.onCreate(QuickstepLauncher.java:527)
	at app.lawnchair.LawnchairLauncher.onCreate(LawnchairLauncher.kt:208)
	at android.app.Activity.performCreate(Activity.java:7159)
	at android.app.Activity.performCreate(Activity.java:7150)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1272)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3001)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3156)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1864)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loop(Looper.java:205)
	at android.app.ActivityThread.main(ActivityThread.java:6993)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:884)

```

Rafael2616 commented:
I'm also running this on my GitHub Actions

Rafael2616 commented:
Related to the above error:

It looks like 'getSourceLayoutResId' It didn't work below 10, analyzing the code I saw that this has already been addressed elsewhere but not here

https://github.com/LawnchairLauncher/lawnchair/blob/13-dev/src%2Fcom%2Fandroid%2Flauncher3%2FWorkspace.java

Rafael2616 commented:
Can I simply check what the version is and set a value if it is below 10? If yes, what value, false or true?

Rafael2616 committed:
fix issue below

Rafael2616 committed:
Update Workspace.java

Rafael2616 commented:
New problem, resource not found 

`android.R.color.system_accent3_200 `

```
Lawnchair (Debug) bug report 2 de dez de 2023 23:12:32
version: 13 Dev (#12) (13)
commit: efc4841
build.brand: lge
build.device: mmh4
build.display: PKQ1.190522.001
build.fingerprint: lge/mmh4_f/mmh4:9/PKQ1.190522.001/2106722188f70:user/release-keys
build.hardware: mmh4
build.id: PKQ1.190522.001
build.manufacturer: LGE
build.model: LM-X420
build.security.level: 2021-03-01
build.product: mmh4_f
build.type: user
version.codename: REL
version.incremental: 2106722188f70
version.release: 9
version.sdk_int: 28
display.density_dpi: 280
isRecentsEnabled: false

error: Uncaught exception

android.content.res.Resources$NotFoundException: Resource ID #0x1060055
	at android.content.res.ResourcesImpl.getValue(ResourcesImpl.java:230)
	at android.content.res.Resources.getColor(Resources.java:964)
	at com.android.launcher3.BubbleTextView.applyIconAndLabel(BubbleTextView.java:396)
	at com.android.launcher3.BubbleTextView.applyFromWorkspaceItem(BubbleTextView.java:347)
	at com.android.launcher3.BubbleTextView.applyFromWorkspaceItem(BubbleTextView.java:317)
	at com.android.launcher3.BubbleTextView.applyFromWorkspaceItem(BubbleTextView.java:312)
	at com.android.launcher3.Launcher.createShortcut(Launcher.java:1430)
	at com.android.launcher3.Launcher.createShortcut(Launcher.java:1415)
	at com.android.launcher3.Launcher.bindItems(Launcher.java:2525)
	at com.android.launcher3.Launcher.bindItems(Launcher.java:2486)
	at com.android.launcher3.model.BaseLauncherBinder$UnifiedWorkspaceBinder.lambda$bindWorkspaceItems$7(BaseLauncherBinder.java:331)
	at com.android.launcher3.model.BaseLauncherBinder$UnifiedWorkspaceBinder$$ExternalSyntheticLambda10.execute(Unknown Source:6)
	at com.android.launcher3.model.BaseLauncherBinder$UnifiedWorkspaceBinder.lambda$executeCallbacksTask$9(BaseLauncherBinder.java:352)
	at com.android.launcher3.model.BaseLauncherBinder$UnifiedWorkspaceBinder.$r8$lambda$a4Qxm57bpz83-gfmBrdCLjV8bf4(Unknown Source:0)
	at com.android.launcher3.model.BaseLauncherBinder$UnifiedWorkspaceBinder$$ExternalSyntheticLambda9.run(Unknown Source:4)
	at android.os.Handler.handleCallback(Handler.java:873)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loop(Looper.java:205)
	at android.app.ActivityThread.main(ActivityThread.java:6993)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:884)

```

Rafael2616 committed:
fix issue below

Rafael2616 committed:
Update BubbleTextView.java

Rafael2616 commented:
Ok, this is the newest problem and I don't know how to solve it
The method 'viewRootImpl.getSurfaceControl()' It's not available on Android 9 and I don't know how to replace it here 

```
private void onLauncherDraw() {
        View view = mLauncher.getDragLayer();
        ViewRootImpl viewRootImpl = view.getViewRootImpl();
        setSurface(viewRootImpl != null ? viewRootImpl.getSurfaceControl() : null);
        view.post(() -> view.getViewTreeObserver().removeOnDrawListener(mOnDrawListener));
    }

```

Rafael2616 committed:
try to fix issue in DepthController.java

I'm not sure how this solution behaves, I'll just compile it to see

Rafael2616 committed:
Update DepthController.java

Rafael2616 commented:
I did it, now it works 🎉🎉😅

Rafael2616 commented:
![Screenshot_20231203-011441.png](https://github.com/LawnchairLauncher/lawnchair/assets/93414086/edd75789-79a0-4ae0-8477-345ae46f0504)

Rafael2616 commented:
![Screenshot_20231203-011448.png](https://github.com/LawnchairLauncher/lawnchair/assets/93414086/0a06d8ce-bddd-485b-a473-1e5e2f6099fe)

Rafael2616 commented:
No very obvious crashes so far, in fact, no crashes at all, just the Doxk search window that is full. But it works

Rafael2616 committed:
Update Workspace.java

MrSluffy commented on quickstep/src/com/android/launcher3/statehandlers/DepthController.java:
```diff
@@ -174,4 +185,4 @@ public void dump(String prefix, PrintWriter writer) {
     }
-}
\ No newline at end of file
+}
```
Clean up, don't add unneccsary spacing

MrSluffy commented on quickstep/src/com/android/quickstep/RecentsAnimationDeviceState.java:
```diff
@@ -208,6 +210,7 @@ public void onExclusionChanged(Region region) {
         } catch (RemoteException e) {
             // Do nothing
         }
+        
```
this too

MrSluffy commented on quickstep/src/com/android/quickstep/RecentsAnimationDeviceState.java:
```diff
@@ -258,6 +263,10 @@ public void onDisplayInfoChanged(Context context, Info info, int flags) {
             mMode = info.navigationMode;
             mNavBarPosition = new NavBarPosition(mMode, info);
 
+            if (mExclusionListener == null) {
```
`if (mExclusionListener == null) return`

MrSluffy commented on src/com/android/launcher3/BubbleTextView.java:
```diff
@@ -389,11 +389,17 @@ protected void setItemInfo(ItemInfoWithIcon itemInfo) {
         mDotParams.appColor = iconDrawable.getIconColor();
         mDotParams.color = getContext().getResources()
-                .getColor(android.R.color.system_accent3_200, getContext().getTheme());
+                .getColor(mColor, getContext().getTheme());
```
```
mDotParams.color = iconDrawable.getIconColor();
```

Rafael2616 replied:
This caused the app to crash again
```
Lawnchair (Debug) bug report 3 de dez de 2023 02:33:03
version: 13 Dev (#22) (13)
commit: 7c3c8d5
build.brand: lge
build.device: mmh4
build.display: PKQ1.190522.001
build.fingerprint: lge/mmh4_f/mmh4:9/PKQ1.190522.001/2106722188f70:user/release-keys
build.hardware: mmh4
build.id: PKQ1.190522.001
build.manufacturer: LGE
build.model: LM-X420
build.security.level: 2021-03-01
build.product: mmh4_f
build.type: user
version.codename: REL
version.incremental: 2106722188f70
version.release: 9
version.sdk_int: 28
display.density_dpi: 280
isRecentsEnabled: false

error: Uncaught exception

android.content.res.Resources$NotFoundException: Resource ID #0xff8ab1e7
	at android.content.res.ResourcesImpl.getValue(ResourcesImpl.java:230)
	at android.content.res.Resources.getColor(Resources.java:964)
	at com.android.launcher3.BubbleTextView.applyIconAndLabel(BubbleTextView.java:396)
	at com.android.launcher3.BubbleTextView.applyFromWorkspaceItem(BubbleTextView.java:347)
	at com.android.launcher3.BubbleTextView.applyFromWorkspaceItem(BubbleTextView.java:317)
	at com.android.launcher3.BubbleTextView.applyFromWorkspaceItem(BubbleTextView.java:312)
	at com.android.launcher3.Launcher.createShortcut(Launcher.java:1430)
	at com.android.launcher3.Launcher.createShortcut(Launcher.java:1415)
	at com.android.launcher3.Launcher.bindItems(Launcher.java:2525)
	at com.android.launcher3.Launcher.bindItems(Launcher.java:2486)
	at com.android.launcher3.model.BaseLauncherBinder$UnifiedWorkspaceBinder.lambda$bindWorkspaceItems$7(BaseLauncherBinder.java:331)
	at com.android.launcher3.model.BaseLauncherBinder$UnifiedWorkspaceBinder$$ExternalSyntheticLambda10.execute(Unknown Source:6)
	at com.android.launcher3.model.BaseLauncherBinder$UnifiedWorkspaceBinder.lambda$executeCallbacksTask$9(BaseLauncherBinder.java:352)
	at com.android.launcher3.model.BaseLauncherBinder$UnifiedWorkspaceBinder.$r8$lambda$a4Qxm57bpz83-gfmBrdCLjV8bf4(Unknown Source:0)
	at com.android.launcher3.model.BaseLauncherBinder$UnifiedWorkspaceBinder$$ExternalSyntheticLambda9.run(Unknown Source:4)
	at android.os.Handler.handleCallback(Handler.java:873)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loop(Looper.java:205)
	at android.app.ActivityThread.main(ActivityThread.java:6993)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:884)

```

MrSluffy commented on src/com/android/launcher3/BubbleTextView.java:
```diff
@@ -389,11 +389,17 @@ protected void setItemInfo(ItemInfoWithIcon itemInfo) {
     @UiThread
     protected void applyIconAndLabel(ItemInfoWithIcon info) {
+        int mColor;
+        if (Utilities.ATLEAST_Q) {
```
remove this

MrSluffy commented on quickstep/src/com/android/quickstep/RecentsAnimationDeviceState.java:
```diff
@@ -216,12 +219,14 @@ public void onActivityPinned(String packageName, int userId, int taskId, int sta
             @Override
             public void onActivityUnpinned() {
-                mPipIsActive = false;
+               mPipIsActive = false;
```
this too

MrSluffy commented on quickstep/src/com/android/launcher3/statehandlers/DepthController.java:
```diff
@@ -64,7 +67,15 @@ public DepthController(Launcher l) {
+        if (Utilities.ATLEAST_Q) {
+            setSurface(viewRootImpl != null ? viewRootImpl.getSurfaceControl() : null);
+        } else {
+            Bitmap bitmap = view.getDrawingCache();
```
why do we need this? exactly?

Rafael2616 replied:
In fact it was the equivalent replacement that I found for the "setSurface" function with the method that is incompatible, I don't know if just leaving it without it will work.

Rafael2616 replied:
> why do we need this? exactly?

And i also need to convert Bitmap to Canvas

Rafael2616 committed:
Update DepthController.java

Rafael2616 committed:
Update RecentsAnimationDeviceState.java

Rafael2616 committed:
Update RecentsAnimationDeviceState.java

Rafael2616 committed:
Update BubbleTextView.java

Rafael2616 committed:
Update RecentsAnimationDeviceState.java

Goooler committed:
Tweak CI configs (#3773)

* Tweak CI config

- Verify release build for every commit or PR, so that we can found potential obfuscation issues.
- Merge build_debug_apk.yml with build_release_apk.yml.
- Rename build_debug_apk.yml to ci.yml

* Cleanup

* Update ci.yml

* Fix release_update.yml

* Revert

MrSluffy committed:
Update fixes

MrSluffy committed:
Update DepthController.java

MrSluffy committed:
Update Workspace.java

suphon-t commented on src/com/android/launcher3/BubbleTextView.java:
```diff
@@ -392,8 +392,7 @@ protected void applyIconAndLabel(ItemInfoWithIcon info) {
         mDotParams.appColor = iconDrawable.getIconColor();
-        mDotParams.color = getContext().getResources()
-                .getColor(android.R.color.system_accent3_200, getContext().getTheme());
+        mDotParams.color = iconDrawable.getIconColor();
```
Lawnchair has a color theming system for these dynamic colors. Please add accent3_200 [here](https://github.com/LawnchairLauncher/lawnchair/blob/13-dev/lawnchair/src/app/lawnchair/theme/color/ColorTokens.kt) and use it instead.

Rafael2616 replied:
@MrSlufffy Could you do this?

suphon-t commented on lawnchair/res/values-v28/colors.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <color name="system_accent1_100">@android:color/transparent</color>
```
Please use something more appropriate than transparent. This can cause visual glitches or even crashes in some places.

Rafael2616 replied:
What color would you think is appropriate as I'm just using the same solution as was used to fix crash on Android 10, 11 by @MrSluffy

suphon-t replied:
Let's leave it like this for now if you say it works

renovate[bot] committed:
Update plugin com.diffplug.spotless to v6.23.3 (#3821)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

validcube committed:
Update repository badges (#3822)

renovate[bot] committed:
Update plugin com.gradle.enterprise to v3.16 (#3824)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

Rafael2616 committed:
Merge branch 'LawnchairLauncher:13-dev' into Rafael2616-patch-1

Rafael2616 committed:
add Accent3_200 in ColorTokens.kt

MrSluffy committed:
Revert "Merge branch 'LawnchairLauncher:13-dev' into Rafael2616-patch-1"

This reverts commit f9fb892e069880ec103944518ff0fc641ee6288b, reversing
changes made to 1c918aa17247d6f5ebd152320c1c96daffc42461.

MrSluffy commented:
I revert the merged as it messes with the commit change

Rafael2616 committed:
Update BubbleTextView.java

Rafael2616 commented:
Ok

Rafael2616 commented:
Could you see that

https://github.com/LawnchairLauncher/lawnchair/pull/3808#discussion_r1412912415

Rafael2616 commented:
How to solve this after applying @suphon-t suggestion

/home/runner/work/lawnchair/lawnchair/src/com/android/launcher3/BubbleTextView.java:396: error: Accent3_200 has private access in ColorTokens

MrSluffy committed:
Fixed error build

Rafael2616 commented:
Thanks

Rafael2616 commented:
It still works 👍. it needs only my 2 review comments above so it's ready to merge

Rafael2616 committed:
fix NoClassDefFoundError 'ActivityTaskManager'

```
Lawnchair (Debug) bug report 6 de dez de 2023 13:43:28
version: 13 Dev (#27) (13)
commit: 83bc0e7
build.brand: lge
build.device: mmh4
build.display: PKQ1.190522.001
build.fingerprint: lge/mmh4_f/mmh4:9/PKQ1.190522.001/2106722188f70:user/release-keys
build.hardware: mmh4
build.id: PKQ1.190522.001
build.manufacturer: LGE
build.model: LM-X420
build.security.level: 2021-03-01
build.product: mmh4_f
build.type: user
version.codename: REL
version.incremental: 2106722188f70
version.release: 9
version.sdk_int: 28
display.density_dpi: 280
isRecentsEnabled: false

error: Uncaught exception

java.lang.NoClassDefFoundError: Failed resolution of: Landroid/app/ActivityTaskManager;
	at com.android.systemui.shared.system.ActivityManagerWrapper.<init>(ActivityManagerWrapper.java:73)
	at com.android.systemui.shared.system.ActivityManagerWrapper.<clinit>(ActivityManagerWrapper.java:64)
	at com.android.systemui.shared.system.ActivityManagerWrapper.getInstance(ActivityManagerWrapper.java:77)
	at com.android.launcher3.uioverrides.QuickstepLauncher.lambda$onUiChangedWhileSleeping$4(QuickstepLauncher.java:641)
	at com.android.launcher3.uioverrides.QuickstepLauncher.$r8$lambda$ZVdAVIwrbYvUMuRD15SAXpfHLb4(Unknown Source:0)
	at com.android.launcher3.uioverrides.QuickstepLauncher$$ExternalSyntheticLambda2.run(Unknown Source:2)
	at android.os.Handler.handleCallback(Handler.java:873)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loop(Looper.java:205)
	at android.os.HandlerThread.run(HandlerThread.java:65)
Caused by: java.lang.ClassNotFoundException: Didn't find class "android.app.ActivityTaskManager" on path: DexPathList[[zip file "/data/app/app.lawnchair.debug-H8HgEn3LbWYzNB49xdhOng==/base.apk"],nativeLibraryDirectories=[/data/app/app.lawnchair.debug-H8HgEn3LbWYzNB49xdhOng==/lib/arm, /data/app/app.lawnchair.debug-H8HgEn3LbWYzNB49xdhOng==/base.apk!/lib/armeabi-v7a, /system/lib, /system/product/lib]]
	at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:134)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
	... 10 more

```

Rafael2616 commented:
I fixed this problem more, could you check 

https://github.com/LawnchairLauncher/lawnchair/assets/93414086/f2ee6219-f450-4453-bbd9-5672003930f0

Rafael2616 committed:
Update ActivityManagerWrapper.java

Rafael2616 committed:
fix line at the end of the file, I think this is Github's fault 😂

Rafael2616 committed:
Update ActivityManagerWrapper.java

Rafael2616 committed:
fix build

MrSluffy commented on systemUIShared/src/com/android/systemui/shared/system/ActivityManagerWrapper.java:
```diff
@@ -70,8 +73,10 @@ public class ActivityManagerWrapper {
-    private final ActivityTaskManager mAtm = ActivityTaskManager.getInstance();
-    private ActivityManagerWrapper() { }
+    private final ActivityTaskManager mAtm;
+    private ActivityManagerWrapper() {
```
this will cause issue in upper version

Rafael2616 replied:
Okay, we don't want that to happen 😂. What must be done then to make it work in both?

MrSluffy commented on systemUIShared/src/com/android/systemui/shared/system/ActivityManagerWrapper.java:
```diff
@@ -70,8 +73,10 @@ public class ActivityManagerWrapper {
-    private ActivityManagerWrapper() { }
+    private final ActivityTaskManager mAtm;
+    private ActivityManagerWrapper() {
+        mAtm = atleastQ ? ActivityTaskManager.getInstance() : null;
```
this too

MrSluffy commented on systemUIShared/src/com/android/systemui/shared/system/ActivityManagerWrapper.java:
```diff
@@ -345,10 +352,16 @@ public boolean showVoiceSession(IBinder token, Bundle args, int flags) {
-                && (context.getPackageManager().hasSystemFeature(
-                PackageManager.FEATURE_FREEFORM_WINDOW_MANAGEMENT)
-                || freeformDevOption);
+        if (atleastQ) {
```
this too

Rafael2616 commented:
@MrSluffy Could I just revert the last change so this can be merged and then fix this in a separate PR since it doesn't significantly affect usage? So much so that I only noticed it after 5 days of use. Or do you intend to correct it soon?

MrSluffy commented:
> @MrSluffy Could I just revert the last change so this can be merged and then fix this in a separate PR since it doesn't significantly affect usage? So much so that I only noticed it after 5 days of use. Or do you intend to correct it soon?

sure, I'll check it later

Rafael2616 committed:
Revert changes in ActivityManagerWrapper.java

Rafael2616 commented:
I'll create an issue for this specific issue soon

MrSluffy commented:
I think the crash in `java.lang.NoClassDefFoundError: Failed resolution of: Landroid/app/ActivityTaskManager;` are similar with this 282edc72a65057821d65fc303e4e92ad99a3fec3 , it should work once merge

Rafael2616 commented:
> I think the crash in `java.lang.NoClassDefFoundError: Failed resolution of: Landroid/app/ActivityTaskManager;` are similar with this 282edc72a65057821d65fc303e4e92ad99a3fec3 , it should work once merge

Yes, this problem no longer occurred after the merge 👍
